framework/src/constraints/ADPenaltyPeriodicSegmentalConstraint.C

```diff
@@ -0,0 +1,142 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADPenaltyPeriodicSegmentalConstraint.h"
+
+registerMooseObject("MooseApp", ADPenaltyPeriodicSegmentalConstraint);
+
+InputParameters
+ADPenaltyPeriodicSegmentalConstraint::validParams()
+{
+  InputParameters params = ADMortarScalarBase::validParams();
+  params.addClassDescription(
+      "ADPenaltyPeriodicSegmentalConstraint enforces macro-micro periodic conditions between "
+      "secondary and primary sides of a mortar interface using a penalty approach "
+      "(no Lagrange multipliers needed). Must be used alongside PenaltyEqualValueConstraint.");
+  params.renameCoupledVar("scalar_variable", "epsilon", "Primary coupled scalar variable");
+  params.addRequiredCoupledVar("sigma", "Controlled scalar averaging variable");
+  params.addParam<Real>(
+      "penalty_value",
+      1.0,
+      "Penalty value used to impose a generalized force capturing the mortar constraint equation");
+
+  return params;
+}
+
+ADPenaltyPeriodicSegmentalConstraint::ADPenaltyPeriodicSegmentalConstraint(
+    const InputParameters & parameters)
+  : DerivativeMaterialInterface<ADMortarScalarBase>(parameters),
+    _temp_jump_global(),
+    _tau_s(),
+    _kappa_aux_ptr(getScalarVar("sigma", 0)),
+    _ka_order(_kappa_aux_ptr->order()),
+    _kappa_aux(coupledScalarValue("sigma")),
+    _pen_scale(getParam<Real>("penalty_value"))
+{
+  if (_kappa_aux_ptr->kind() != Moose::VarKindType::VAR_AUXILIARY)
+    mooseError("Must assign auxiliary scalar variable to sigma, rather than nonlinear variable");
```

Comment:
Love it!

Reply:
Oh wait
```suggestion
  if (_kappa_aux_ptr->kind() != Moose::VarKindType::VAR_AUXILIARY)
    paramError("sigma", "Must assign auxiliary scalar variable to sigma, rather than nonlinear variable");
```
Try to use `paramError` when there's an issue with a provided parameter/coupled-variable